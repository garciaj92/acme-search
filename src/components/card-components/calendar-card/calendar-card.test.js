import React from "react";
import CalendarCard from "../calendar-card";

import { mount, shallow } from "enzyme";
import toJson from 'enzyme-to-json';

// Mocks
import { consoleErrorFn } from "../../../__mocks__/console.mock";

const props = {
    title: "Calendar Test",
    invitees: ["Test Guest"],
    date: "01/01/2020 - 10:00:00 AM"
};

describe("CalendarCard: Test visual components", () => {
    let wrapper;
    beforeEach(() => {
        wrapper = mount(<CalendarCard {...props} />);
    });

    it("renders correctly", () => {
        expect(toJson(wrapper)).toMatchSnapshot();
        // console.log(wrapper.debug());
    });

    it("displays sets the props correctly", () => {
        expect(wrapper.find("[data-testid='card-heading']").first().text()).toEqual(props.title);
        expect(wrapper.find("[data-testid='guest-info'] span").first().text()).toEqual(props.invitees[0]);
    });
});


describe("CalendarCard: Test prop type validation", () => {
    let consoleErrorSpy;
    beforeEach(() => {
        console.error = consoleErrorFn
        consoleErrorSpy = jest.spyOn(console, "error");
    });

    afterEach(() => {
        consoleErrorSpy.mockClear();
    });
    
    it("does not throw an error when props are provided", () => {
        shallow(<CalendarCard {...props} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(0);
    });

    it("does not throw an error when invitees prop is not provided", () => {
        const { 
            title,
            date
        } = props;
        shallow(<CalendarCard title={title} date={date} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(0);
    });

    // What is stated below applies to the remaining test cases in this describe block:
    //      console.error is called twice for each prop validator that fails because
    //      there are two components that use each failing prop. This means that if
    //      two props fail we have four errors.

    it("throws an error when date prop is not provided", () => {
        const { 
            title,
            invitees,
        } = props;
        shallow(<CalendarCard title={title} invitees={invitees} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(2);
    });

    it("throws an error when title prop is not provided", () => {
        const { 
            invitees,
            date
        } = props;
        shallow(<CalendarCard date={date} invitees={invitees} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(2);
    });

    it("throws an error when either title or date prop are not of type `string`", () => {
        shallow(<CalendarCard title={0} date={0}  />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(4);
    });

    it("throws an error when invitees prop is not an array of type `string`", () => {
        const { 
            title,
            date
        } = props;
        shallow(<CalendarCard title={title} invitees={[0,2]} date={date} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(2);
    });
});
