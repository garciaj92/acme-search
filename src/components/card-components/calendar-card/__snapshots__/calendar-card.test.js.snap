// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarCard: Test visual components renders correctly 1`] = `
<CalendarCard
  date="01/01/2020 - 10:00:00 AM"
  invitees={
    Array [
      "Test Guest",
    ]
  }
  title="Calendar Test"
>
  <Card
    bg="white"
    body={false}
    className="shadow calendar-card"
  >
    <div
      className="shadow calendar-card card bg-white"
    >
      <CardBody>
        <div
          className="card-body"
        >
          <CardHeading
            data-testid="card-heading"
            heading="Calendar Test"
          >
            <h4
              className="font-weight-bold"
            >
              Calendar Test
            </h4>
          </CardHeading>
          <GuestsInfo
            data-testid="guest-info"
            invitees={
              Array [
                "Test Guest",
              ]
            }
          >
            <div
              className="d-flex justify-content-center align-items-center my-2"
            >
              <Image
                fluid={false}
                rounded={false}
                roundedCircle={false}
                src="people-outline.svg"
                thumbnail={false}
                width={30}
              >
                <img
                  className=""
                  src="people-outline.svg"
                  width={30}
                />
              </Image>
              <h5
                className="mb-0 ml-1"
              >
                1
                 invited
              </h5>
            </div>
            <div
              className="d-flex flex-wrap justify-content-center"
            >
              <Badge
                className="m-1"
                key="0"
                pill={true}
                variant="info"
              >
                <span
                  className="m-1 badge badge-pill badge-info"
                >
                  Test Guest
                </span>
              </Badge>
            </div>
          </GuestsInfo>
        </div>
      </CardBody>
      <CardFooter
        className="bg-white"
        data-testid="card-footer"
      >
        <div
          className="bg-white card-footer"
          data-testid="card-footer"
        >
          <DateInfo
            date="01/01/2020 - 10:00:00 AM"
            title="Time & Date"
          >
            <div
              className="text-muted"
            >
              <h5
                className="font-weight-bold mb-0"
              >
                Time & Date
              </h5>
              <p
                className="font-weight-bold mb-0"
              >
                01/01/2020 - 10:00:00 AM
              </p>
            </div>
          </DateInfo>
        </div>
      </CardFooter>
    </div>
  </Card>
</CalendarCard>
`;
